NIFI-14629 Added timeout blocking to Kinesis StandardRecordProcessorB…

dariuszseweryn opened the pull request:
…locker

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-14629](https://issues.apache.org/jira/browse/NIFI-14629)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] (there are no new dependencies) New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] (there are no new dependencies) New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] (there is no new documentation) Documentation formatting appears as expected in rendered files

exceptionfactory commented on nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/pause/StandardRecordProcessorBlocker.java:
```diff
@@ -38,7 +76,17 @@ public synchronized void block() {
+        unblockPrivate(Long.MAX_VALUE);
+    }
+
+    private synchronized void unblockPrivate(final long newTimeoutAnchor) {
```
Use of `Private` in the method name should be avoided. `Internal` could work, but better to be more descriptive, such as `countDown()` based on what the method is doing.

exceptionfactory commented on nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/pause/StandardRecordProcessorBlocker.java:
```diff
@@ -17,19 +17,57 @@
 
+    static final long BLOCK_AFTER_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(2);
+    private final Supplier<Long> timestampProvider;
+    private long timeoutAnchor = 0L;
```
The naming of this property is unclear, it should be renamed to something that more clearly describes its purpose.

dariuszseweryn replied:
Renamed to `timeoutBase`

exceptionfactory reviewed (changes requested):
Thanks for evaluating potential improvements to Kinesis processing @dariuszseweryn. On initial review, I noticed several issues.

The introduction of the new `unblockAndDisableTimeout()` method is only on the Standard implementation, which goes outside the bounds of the interface definition.

More to the substance of the implementation, however, the `Blocker` class now includes two CountDownLatch instances, and synchronization at different points, leading to a difficult to follow approach. I think the approach needs to be revisited and simplified, otherwise it is introducing more complexity to something that is already getting too complex to be maintainable. Some of this relates to fundamental design issues with the Processor itself, and attempting to work around them is going to be a losing battle in the end.

dariuszseweryn commented:
> The introduction of the new unblockAndDisableTimeout() method is only on the Standard implementation, which goes outside the bounds of the interface definition.

The interface defines only `await()` method — `block()`, `unblock()` are only on the Standard implementation — this was the accepted solution in the prior PR. `unblockAndDisableTimeout()` only builds on that approach. What you propose?

I have extracted `TimeoutActivationLatch` class and changed naming. Hopefully it is now clearer. PTAL

dariuszseweryn commented:
> From an implementation point of view, it would be helpful to make use of java.time.Duration in several places, versus the separate TimeUnit and long milliseconds.

Changed to used Duration. I'll use it more often going forward.
> As far as the Supplier approach for the timestamp, is that necessary? It adds a layer of indirection that makes the implementation more difficult to follow, although I can see where it might be useful only for testing.

Yes, it helps for testing, that is the purpose. Public interface is oblivious of the Supplier though.
> Regarding testing, is the Noop implementation added to avoid timing issues with the tests?

Correct. I have changed to use the standard implementation and call `unblockInfinitely()` on it.
> One more high-level question, is there a reason for maintaining both the CountDownLatch and the Timeout Latch, as opposed to just the Timeout Latch? This still seems like some more fundamental that needs to be reconsidered.

I wanted to have a separation of concerns, but since I have merged TimeoutActivationLatch into StandardRecordProcessorBlocker I am now using a single latch and change behavior by changing `timeoutBase` parameter.

exceptionfactory reviewed (approved):
Thanks for making the adjustments @dariuszseweryn, the latest version is much more streamlined, and looks like a reasonable approach given the current implementation constraints with the Kinesis library.
